### PR TITLE
Test framework API versioning changes

### PR DESCRIPTION
**Pull request by miquelsi**
https://redhat.atlassian.net/browse/ARO-27211

### What

- Added API version abstraction in framework
- The default API version is the oldest one (20240610). It can be configured using `ARO_HCP_E2E_API_VERSION` environment variable, or specifically for one test with `tc.SetHCPAPIVersionForTest(framework.HCPAPIVersion*)`
- Refactored cluster and node pool creation routing
- Added facade for cluster and node pool clients
- Added version-agnostic helpers for admin credential/revoke  helpers, nodepool disk-type validations and KeyVault visibility checks

### Why

- Decouple tests from SDK version churn: tests no longer need to hardcode which API client/helper to use for common operations.
- Centralize version selection: one place (env or test context) controls API version behavior.
- Safer migration path: wrappers preserve compatibility while enabling gradual cleanup.
- Cleaner test code: fewer low-level SDK calls in tests; framework owns version routing.
- Future-ready: adding a new API version mostly becomes framework wiring, not mass test rewrites.

### PR Checklist
- [x] PR is scoped to a single task (no mixed concerns)
- [x] Title follows [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/) format
- [x] Summary explains the "Why" behind the change
- [x] Linked to relevant ticket/issue
- [ ] Screenshots included (if graph/UI/metrics changes)
- [ ] Self-reviewed the diff
- [x] CI/CD checks are passing (ignore Tide)
- [x] Draft PR used for WIP (if applicable)
- [x] Commit history is clean (rebased/squashed)
- [ ] Tricky code blocks are commented
- [ ] Specific reviewers tagged
- [ ] All comment threads resolved before merge

### CODE REVIEW

File: test/util/framework/per_test_framework.go [hunk 1]
```diff
@@ -1090,6 +1125,27 @@ func (tc *perItOrDescribeTestContext) TenantID() string {
 	return tc.perBinaryInvocationTestContext.tenantID
 }
 
+func (tc *perItOrDescribeTestContext) SetHCPAPIVersionForTest(version HCPAPIVersion) {
+	tc.contextLock.Lock()
+	defer tc.contextLock.Unlock()
+	tc.hcpAPIVersionOverride = &version
+}
```
Comment (mbukatov) [inline test/util/framework/per_test_framework.go:1132]: Could we log what version was selected? That will allow us to filter test results using particular API version if needed.

File: test/util/framework/per_test_framework.go [hunk 2]
```diff
@@ -1090,6 +1125,27 @@ func (tc *perItOrDescribeTestContext) TenantID() string {
 	return tc.perBinaryInvocationTestContext.tenantID
 }
 
+func (tc *perItOrDescribeTestContext) SetHCPAPIVersionForTest(version HCPAPIVersion) {
+	tc.contextLock.Lock()
+	defer tc.contextLock.Unlock()
```
Comment (mbukatov) [inline test/util/framework/per_test_framework.go:1130]: I have one concern with this approach: it will make a test which deploys multiple hosted clusters using different APIs impossible. That said, I'm not sure how bit a problem this is.

File: test/e2e/cluster_create_complex_cilium_kv.go [hunk 3]
```diff
@@ -50,6 +45,7 @@ var _ = Describe("Customer", func() {
 			)
 
 			tc := framework.NewTestContext()
+			tc.SetHCPAPIVersionForTest(framework.HCPAPIVersion20251223)
```
Comment (mbukatov) [inline test/e2e/cluster_create_complex_cilium_kv.go:48]: So basically a test case will use default API version, unless a different API version is selected using `tc.SetHCPAPIVersionForTest()`, right?

Personally I would prefer an explicit selection (so that every test would need make the decision explicitly), but this will work as well.

File: test/e2e/cluster_create_complex_cilium_kv.go [hunk 4]
```diff
@@ -81,33 +77,19 @@ var _ = Describe("Customer", func() {
...
-			clusterResource, err := framework.BuildHCPCluster20251223FromParams(clusterParams, tc.Location(), nil)
-			Expect(err).NotTo(HaveOccurred())
-
-			// Set KeyVault visibility to Private
-			if clusterResource.Properties != nil && clusterResource.Properties.Etcd != nil &&
-				clusterResource.Properties.Etcd.DataEncryption != nil &&
-				clusterResource.Properties.Etcd.DataEncryption.CustomerManaged != nil &&
-				clusterResource.Properties.Etcd.DataEncryption.CustomerManaged.Kms != nil {
-				clusterResource.Properties.Etcd.DataEncryption.CustomerManaged.Kms.Visibility = to.Ptr(hcpsdk20251223preview.KeyVaultVisibilityPrivate)
-			}
```
Comment (mbukatov) [inline test/e2e/cluster_create_complex_cilium_kv.go:94]: It seems to me that these lines to reconfigure KeyVault visibility should be preserved.

File: test/e2e/cluster_create_complex_cilium_kv.go [hunk 5]
```diff
@@ -156,55 +138,24 @@ var _ = Describe("Customer", func() {
...
-					Platform: &hcpsdk20251223preview.NodePoolPlatformProfile{
-						VMSize: to.Ptr(nodePoolParams.VMSize),
-						OSDisk: &hcpsdk20251223preview.OsDiskProfile{
-							SizeGiB:                to.Ptr(nodePoolParams.OSDiskSizeGiB),
-							DiskStorageAccountType: to.Ptr(hcpsdk20251223preview.DiskStorageAccountType(nodePoolParams.DiskStorageAccountType)),
-						},
-					},
-					AutoRepair: to.Ptr(true),
-				},
-			}
```
Comment (mbukatov) [inline test/e2e/cluster_create_complex_cilium_kv.go:178]: Here I assume these are all defaults we don't need specify when using NewDefaultNodePoolParams(). But will recheck.

File: test/e2e/cluster_create_complex_cilium_kv.go [hunk 6]
```diff
@@ -156,55 +138,24 @@ var _ = Describe("Customer", func() {
...
 				*resourceGroup.Name,
+				clusterParams.ManagedResourceGroupName,
 				customerClusterName,
-				customerNodePoolName,
-				nodePool,
+				nodePoolParams,
 				45*time.Minute,
 			)
-			// We delay checking the error on purpose to get more details
-			// about the issue by running the verifiers.
```
Comment (mbukatov) [inline test/e2e/cluster_create_complex_cilium_kv.go:190]: I would keep this comment, explaining that the omission of immediate nodePoolErr checking is done on purpose.

File: test/e2e/cluster_create_complex_cilium_kv.go [hunk 7]
```diff
@@ -156,55 +138,24 @@ var _ = Describe("Customer", func() {
...
-						clusterParams.ManagedResourceGroupName,
-						tc.LogDirPath)
-				}
-			}
+			Expect(nodePoolErr).NotTo(HaveOccurred())
 
 			By("verifying nodes become Ready with Cilium CNI")
 			err = verifiers.VerifyHCPCluster(ctx, adminRESTConfig, verifiers.VerifyNodesReady(), verifiers.VerifyCiliumOperational("kube-system", "k8s-app=cilium"))
-			Expect(errors.Join(err, nodePoolErr, consoleLogErr)).NotTo(HaveOccurred())
+			Expect(err).NotTo(HaveOccurred())
```
Comment (mbukatov) [inline test/e2e/cluster_create_complex_cilium_kv.go:158]: This should be: `Expect(errors.Join(err, nodePoolErr)).NotTo(HaveOccurred())`

File: test/e2e/cluster_create_private_kv.go [hunk 8]
```diff
@@ -79,24 +77,11 @@ var _ = Describe("Create HCPOpenShiftCluster with Private KeyVault", func() {
...
-			clusterResource, err := framework.BuildHCPCluster20251223FromParams(clusterParams, tc.Location(), nil)
-			Expect(err).NotTo(HaveOccurred())
-
-			// Set KeyVault visibility
-			if clusterResource.Properties != nil && clusterResource.Properties.Etcd != nil &&
-				clusterResource.Properties.Etcd.DataEncryption != nil &&
-				clusterResource.Properties.Etcd.DataEncryption.CustomerManaged != nil &&
-				clusterResource.Properties.Etcd.DataEncryption.CustomerManaged.Kms != nil {
-				clusterResource.Properties.Etcd.DataEncryption.CustomerManaged.Kms.Visibility = to.Ptr(hcpsdk20251223preview.KeyVaultVisibilityPrivate)
-			}
```
Comment (mbukatov) [inline test/e2e/cluster_create_private_kv.go:91]: Here again, don't we need to make sure that KV visibility is set to private?

File: test/e2e/cluster_create_private_kv.go [hunk 9]
```diff
@@ -108,33 +93,22 @@ var _ = Describe("Create HCPOpenShiftCluster with Private KeyVault", func() {
...
-				customerClusterName,
-				nil,
-			)
+			visibility, err := tc.GetClusterKeyVaultVisibility(ctx, *resourceGroup.Name, customerClusterName)
 			Expect(err).ToNot(HaveOccurred())
-			Expect(cluster.Properties).ToNot(BeNil())
-			Expect(cluster.Properties.Etcd).ToNot(BeNil())
-			Expect(cluster.Properties.Etcd.DataEncryption).ToNot(BeNil())
-			Expect(cluster.Properties.Etcd.DataEncryption.CustomerManaged).ToNot(BeNil())
-			Expect(cluster.Properties.Etcd.DataEncryption.CustomerManaged.Kms).ToNot(BeNil())
```
Comment (mbukatov) [inline test/e2e/cluster_create_private_kv.go:123]: Don't we want to keep this checks?

File: test/util/framework/deployment_helper.go [hunk 10]
```diff
@@ -328,79 +349,110 @@ func (tc *perItOrDescribeTestContext) CreateHCPClusterFromParam(
...
-		ctx, cancel = context.WithTimeoutCause(ctx, timeout, fmt.Errorf("timeout '%f' minutes exceeded during CreateHCPClusterFromParam for cluster %s in resource group %s", timeout.Minutes(), parameters.ClusterName, resourceGroupName))
-		defer cancel()
-	}
-	clusterName := parameters.ClusterName
-
-	startTime := time.Now()
-	defer func() {
-		finishTime := time.Now()
-		tc.RecordTestStep(fmt.Sprintf("Deploy HCP cluster %s/%s", resourceGroupName, clusterName), startTime, finishTime)
-	}()
```
Comment (mbukatov) [inline test/util/framework/deployment_helper.go:342]: Did we drop these test steps on purpose?

File: test/util/framework/hcp_client_facade.go [hunk 11]
```diff
@@ -0,0 +1,467 @@
...
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/http"
+
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore/runtime"
+
+	hcpsdk20240610preview "github.com/Azure/ARO-HCP/test/sdk/resourcemanager/redhatopenshifthcp/armredhatopenshifthcp"
```
Comment (mbukatov) [inline test/util/framework/hcp_client_facade.go:26]: maybe we can use `/test/sdk/v20240610preview/resourcemanager/redhatopenshifthcp/armredhatopenshifthcp` here directly

**Review by mbukatov — COMMENTED** [review]
Thank you for looking into this quickly, this is an initial quick review.

Personally, I would rather avoid a per test global state of selected API version. But this approach would work as well.

Also, I need to think about the suggested interface over the API versions. I would prefer more explicit approach with version visible in name of a function or it's parameter in each API call. That said, I don't have a strong opinion about it right now, and the proposed approach will work as well.

**Review by mbarnes — COMMENTED** [review]
The override mechanism that you've defined (`SetHCPAPIVersionForTest`) is fine for corner cases but I think a label-based approach would be more sustainable, so tests can declare the _minimum_ API version they need instead of locking themselves to one version with the override.  The override defeats the purpose of the abstraction.

Thankfully the way Azure API versions are named, they sort lexicographically over time.  So if you introduced a label for each API version in the pattern of, say, `MinimumVersionYYYYMMDD[preview]`, then you could derive which API versions a test is compatible with by sorting those labels and using whichever one the test declares as the starting point.

Just a suggestion.

**Comment by miquelsi** [comment]
Closing as we won't want to consider API-agnostic tests.
